core/observe.c

```diff
@@ -259,49 +261,158 @@ void cancel_observe(lwm2m_context_t * contextP,
     }
 }
 
+typedef struct {
+  lwm2m_blockwise_t * blockwiseP;
+  char * buffer;
+  int length;
+} notification_data_t;
+
+static void prv_free_notification_data(notification_data_t* data) {
+    if (NULL == data->blockwiseP)
+    {
+        lwm2m_free(data->buffer);
+        data->buffer = NULL;
+    }
+}
+
+coap_status_t lwm2m_sendNotification(lwm2m_context_t * contextP, lwm2m_watcher_t * watcherP, lwm2m_uri_t * uriP, notification_data_t* data) {
+  coap_status_t result = COAP_205_CONTENT;
+  if (NULL == data->buffer) {
+      result = object_read(contextP, uriP, &(data->buffer), &(data->length));
+  }
+  if (result == COAP_205_CONTENT) {
+    coap_packet_t message;
+    watcherP->lastMid = contextP->nextMID++;
+
+    coap_init_message(&message, COAP_TYPE_NON, COAP_205_CONTENT, watcherP->lastMid);
+    coap_set_payload(&message, data->buffer, data->length);
+
+    // if (!LWM2M_URI_IS_SET_RESOURCE(uriP)) {                      //JH+ leshan awaits content-type in case of a object instance notification
+    //   //ToDo: at the moment TLV is default for object instance!  //JH+
+    //   coap_set_header_content_type(message, VND_OMA_LWM2M_TLV);  //JH+ new pre-IANA enums see: er_coap_13.h
+    // }                                                            //JH+
+    coap_set_header_token(&message, watcherP->token, watcherP->tokenLen);
+    coap_set_header_observe(&message, watcherP->counter++);
+    if (watcherP->blockSize < data->length)
+    {
+        if (NULL == data->blockwiseP)
+        {
+            data->blockwiseP = blockwise_new(contextP, uriP, &message, false);
+        }
+        blockwise_prepare(data->blockwiseP, 0, watcherP->blockSize, &message);
+    }
+    LOG("send notify %d %d/%d/%d mid %d\n", watcherP->server->shortID, uriP->objectId, uriP->instanceId, uriP->resourceId, message.mid);
+
+    (void)message_send(contextP, &message, watcherP->server->sessionH);
+  }
+  return result;
+}
+
+time_t lwm2m_notify(lwm2m_context_t * contextP, struct timeval * tv) {
+    lwm2m_observed_t * observedP;
+    time_t nextTransmission = 0;
+    
+    observedP = contextP->observedList;
+
+    while (observedP != NULL)
+    {
+        notification_data_t data;
+        memset(&data, 0, sizeof(data));
+        lwm2m_watcher_t * watcherP = observedP->watcherList;
+        while (watcherP != NULL) {
+          lwm2m_attribute_data_t * attributeData = lwm2m_getAttributes(watcherP->server, &(observedP->uri));
```

Comment:
Maybe you should add in the lwm2m_watcher_t a pointer to the revelant lwm2m_attribute_data_t. This would avoid parsing the list each time.